src/nnet3/nnet-am-decodable-simple.cc

```diff
@@ -203,15 +196,83 @@ void NnetDecodableBase::GetCurrentIvector(int32 output_t_start,
                 << " (mismatched --ivector-period?)";
     }
     ivector_frame = online_ivector_feats_->NumRows() - 1;
+  } else if (ivector_frame < 0) {
+    int32 margin = - ivector_frame;
+    if (margin * online_ivector_period_ > 50) {
+      // Half a second seems like too long to be explainable as edge effects.
+      KALDI_ERR << "Could not get iVector for frame " << frame_to_search
+                << ", only available till frame "
+                << "0 * ivector-period=" << online_ivector_period_
+                << " (mismatched --ivector-period?)";
+    }
+    ivector_frame = 0;
   }
-  *ivector = online_ivector_feats_->Row(ivector_frame);
+  ivector->CopyFromVec(online_ivector_feats_->Row(ivector_frame));
 }
 
+void NnetDecodableBase::GetIvectorsForFrames(int32 output_t_start,
+                                             int32 num_output_frames,
+                                             Matrix<BaseFloat> *ivectors) {
+  const int32 left_context = nnet_left_context_ + opts_.extra_left_context,
+        right_context = nnet_right_context_ + opts_.extra_right_context;
+  // frame_to_search is the frame that we want to get the most recent iVector
+  // for.  We choose a point near the middle of the current window, the concept
+  // being that this is the fairest comparison to nnet2.   Obviously we could do
+  // better by always taking the last frame's iVector, but decoding with
+  // 'online' ivectors is only really a mechanism to simulate online operation.
+  if (ivector_interval_ == 0) { // single ivector case
+    ivectors->Resize(1, online_ivector_feats_->NumCols(), kUndefined);
+    int32 frame_to_search = output_t_start + num_output_frames / 2;
+    SubVector<BaseFloat> sub_ivector = ivectors->Row(0);
+    GetCurrentIvector(frame_to_search, &sub_ivector);
+  } else { // multiple ivectors case
+    // num_ivectors_left is the num of ivectors for frames whose "t" index < 0.
```

Comment:
you still appear to have this strange distinction between the left and right iVectors.
There is no asymmetry.  Try to explain to Vijay why you think there need to be two separate cases and maybe the confusion will become clear.  It's just an interval of frames [output_t_start - left-context ... output_t_start + num_output_frames + right_context), there is no internal structure at all.